packages/gatsby/src/schema/infer/inference-metadata.js

```diff
@@ -216,15 +220,20 @@ const _updateValueDescriptor = (
       })
       break
     }
-    case `listOfUnion`: {
+    case `relatedNode`:
+    case `relatedNodeList`: {
       const { nodes = {} } = typeInfo
-      value.forEach(nodeId => {
+      const listOfNodeIds = Array.isArray(value) ? value : [value]
+      listOfNodeIds.forEach(nodeId => {
         nodes[nodeId] = (nodes[nodeId] || 0) + delta
```

Comment:
Is there value to setting this after `dirty === true`? Otherwise we can use a `some` for this loop so it can short circuit the loop on the first dirty.

(I kind of doubt it since this seems to be a lot of work for just tracking dirty, but who knows)

Reply:
We use this information for incremental inference. Dirty tracking is orthogonal to type inference - it just helps us detect when structural changes occur but we still need to track all entries of all nodes.